동기화 메모장 [STEP 2] 예거, 차차

ChaminLee opened the pull request:
안녕하세요 cory! @corykim0829 
동기화 메모장 STEP 2 구현 완료하여 PR 보내드립니다! 

Core Data의 효율적인 활용과 전체적인 객체들간의 구조를 고민하며 프로젝트를 진행해봤습니다!
내용이 많이 길지만... 이번 스텝도 잘 부탁드립니다 😄 

---

# 🔥 고민했던 점

### 1️⃣ 메모 삭제에 따른 selectedIndexPath 이동 
`UITableView`의 swipeAction 혹은 더보기 버튼을 선택하여 메모를 삭제했을 때 현재 보여질 메모를 나타내기 위한 selectedIndexPath를 업데이트하기 위한 로직을 고민해봤습니다.

- selectedIndexPath 보다 앞의 indexPath에 해당하는 메모를 지우는 경우
    - 현재 보여지는 화면이 유지되어야 하기 때문에 selectedIndexPath의 row를 1만큼 빼주고 업데이트
- selectedIndexPath에 해당하는 메모를 지우는 경우 
    - 데이터의 개수와 비교하여 마지막에 해당한다면 row를 1만큼 빼주고 업데이트
    - 데이터의 개수와 비교하여 마지막이 아니라면 indexPath를 유지하고 화면만 업데이트
- selectedIndexPath 보다 뒤의 indexPath에 해당하는 메모를 지우는 경우
    - 기존의 화면이 보이도록 selectedIndexPath 유지 

위의 로직을 구성하여 코드에 반영하여, 사용자가 메모를 삭제했을 때 자연스럽게 주변 메모로 이동하여 보여줄 수 있도록 구현했습니다. 

### 2️⃣ CoreData CRUD 구현 

CoreData를 관리하는 `MemoStorage`라는 타입을 별도로 생성하여 내부에 CRUD를 각각 구현해두었습니다. `MemoStorage`를 소유하는 `MemoSplitviewController`가 CRUD를 활용하여 전반적인 데이터 관리를 할 수 있도록 도왔습니다.

`MemoStorage` 타입은 Singleton의 단점(테스트가 어렵고, 아무 객체나 쉽게 데이터에 접근하여 변경할 수 있는 등)을 고려하여 구현해주었습니다. 


### 3️⃣ UITextView 변경에 따른 UITableView 화면 동기화 

`UITextView`에 사용자가 텍스트를 입력할 때에 `textViewDidChange()`를 통해 데이터를 업데이트 하도록 해주었습니다.(현재는 core data 접근을 하지 않습니다) 이에 데이터가 변경되었기 때문에 `UITableView` 의 뷰 또한 업데이트가 필요하여 `reloadRows()`를 사용하여 tableView 전체를 업데이트하기 보다는 해당하는 부분만 업데이트 하도록 구현해주었습니다. 

또한 편집을 마치게 되면 `textViewDidEndEditing()` 메서드가 호출되는데, 이 때 편집이 완료된 데이터를 기준으로 먼저 변경이 있는지 여부를 확인해주었습니다. 변경 여부를 확인하기 위해 `textViewDidBeginEditing()`에서 기존 텍스트를 저장하고 비교하는 식으로 구현했습니다. 이에 변경 사항이 있을 때만 core data에 update를 요청하도록 구현해주었습니다. 

텍스트가 변경되는 매 순간 core data에 접근하여 업데이트 하는 것은 비용적으로 비효율적이라고 판단하여 위와 같이 최종 편집 완료시 core data에 업데이트 되도록 구현해주었습니다.

뿐만 아니라, cell의 미리보기 label에 표시 될 때 내용에 줄바꿈이 많은 경우 공백으로 보이는 현상이 있었습니다. 이에 cell의 미리보기 label에 텍스트를 전달해줄 때에는 `trimmingCharacters()`를 사용하여 앞뒤의 줄바꿈을 모두 제거한 상태로 전달해주어 줄바꿈을 무시하고 제목 이후에 등장하는 첫 텍스트를 보여주도록 구현했습니다. 

### 4️⃣ 앱 최초 실행 시 초기 메모를 세팅 

앱을 최초로 실행했을 때에 별도 처리를 하지 않으면 우측의 `UITextView`만 보이게 됩니다. 이에 편집을 제한하지 않으면 텍스트 입력이 가능해지는 데, 이러한 상황을 막기 위해 편집을 제한하기 보다는 최초 1개의 메모를 미리 세팅해두는 방식을 선택했습니다. 

이에 무조건 적으로 메모는 1개 이상 존재하도록 로직을 구현하였습니다. 이 때문에 앱을 최초 실행했을 때는 제목과 내용이 비어있기 때문에 각각 매칭하여 "새로운 메모", "추가 텍스트 없음"을 노출시켜주도록 구현했습니다. 

### 5️⃣ 항상 lastModified를 기준으로 데이터를 정렬

`Memo` 타입의 데이터 배열을 항상 마지막 편집일자인 `lastModified`를 기준으로 내림차순 정렬해주었습니다. 해당 데이터 배열에 접근하여 값을 변경할 때 마다 매번 정렬하도록 property observer를 사용해주었습니다. 

```swift
private var memos = [Memo]() {
    didSet {
        memos.sort { $0.lastModified > $1.lastModified }
    }
}
```

또한 항상 마지막에 편집한 메모가 상단에 노출되도록 하기 위해 편집을 마치면 `moveRow()`를 통해 상단으로 이동하도록 구현해주었습니다. 

### 6️⃣ UITextView내 제목/내용의 폰트를 서로 다르게 구현 

제목과 내용을 육안으로도 구분할 수 있도록 서로 폰트를 다르게 구현해주었습니다. 데이터를 가져와서 `UITextView`에 보여줄 때 `attributedString`을 활용하여 각각 다른 attribute를 가지도록 구현했습니다. 

또한 편집하는 도중에도 첫 번째 줄은 제목이 되고, 그 다음 줄부터는 내용이 되어야 하기 때문에 이를 위해 `textView(_ textView: UITextView, shouldChangeTextIn range: NSRange, replacementText text: String)`를 사용하여 range의 location과 줄바꿈 부호로 구분했을 때 얻을 수 있는 range의 location을 비교하여 첫 번째 줄은 `.largeTitle`, 그 다음부터는 `.title2`로 보여지도록 구현했습니다

```swift
func textView(_ textView: UITextView, shouldChangeTextIn range: NSRange, replacementText text: String) -> Bool {
    let textAsNSString = textView.text as NSString
    let replacedString = textAsNSString.replacingCharacters(in: range, with: text) as NSString
    let titleRange = replacedString.range(of: .lineBreak)

    if titleRange.location > range.location {
        textView.typingAttributes = TextAttribute.title
    } else {
        textView.typingAttributes = TextAttribute.body
    }

    return true
}
```
<img src="https://user-images.githubusercontent.com/45652743/154242847-ad91eab8-4b11-4016-b0fe-75b647d4755b.gif" width=50%> 


### 7️⃣ Delegate Pattern 적용

기존에 `MemoSplitViewController`를 거쳐 `MemoTableViewController`와 `MemoDetailViewController` 간 소통을 도왔던 구조에서 delegate pattern을 적용하였습니다. 

이에 기존에 하위 컨트롤러인 `MemoTableViewController`가 `splitViewController` 프로퍼티를 사용하여 상위 컨트롤러를 알지 못하더라도 delegate를 통해 필요로 하는 기능들을 사용할 수 있도록 구현했습니다. 이를 통해 자식 컨트롤러가 부모 컨트롤러를 아는 부적절한 의존 관계를 제거할 수 있었습니다. 

Delegate pattern 구현을 위해 생성한 프로토콜은 다음과 같습니다. 

- `MemoStorageManageable`: `MemoStorage`의 CRUD를 직접적으로 사용하여 데이터를 관리하는 역할
- `MemoSplitViewManageable` : 전반적인 `UISplitViewController`의 메서드나 하위 뷰컨 간 소통을 위한 역할
- 위 두 프로토콜을 typealias 사용하여 `MemoManageable`를 생성하고 delegate의 타입으로 사용하도록 해주었습니다. 

<br> 

# 🙋🏻‍♂️ 질문하고 싶은 부분

### 1️⃣ CoreData 모델을 커스텀해서 사용해도 될까요?

- 지난 `스텝1`에서는 샘플 JSON 데이터를 사용했었는데요.
`스텝2`로 들어오면서 JSON 대신 `CoreData`를 사용하는 식으로 모델 타입을 변경했습니다. (기존 JSON 파싱 모델 삭제)
이 과정에서 [Create NSManagedObject Subclass] 버튼을 눌러, `CoreDataClass`, `CoreDataProperties` 파일을 만들고 프로젝트 내부로 위치를 옮겼습니다.
- 모델 타입을 만들고 의문을 가졌던 점은, 왜 Entity 에 `Attributes`를 추가할 때 기본적으로 `옵셔널`이 체크되는지 궁금했습니다.
생성된 Attributes 의 옵셔널을 제거하고 편리하게 사용해도 괜찮은지 구글링을 해봤는데요.
Xcode 가 기본적으로 만들어준 속성을 바꾸지 말라는 의견이 좀 더 많이 보이더라구요. 😅 
- 🤚🏻 질문 1 : Entity 의 `Attributes`를 생성할 때 개발자가 원하는 대로 `옵셔널`이나 기타 설정을 조정하고 사용해도 될까요?
사실 id, title, body, lastModified -> 4가지 프로퍼티 모두 옵셔널일 필요는 없는데, 일단은 수정하지 않고, 옵셔널 바인딩을 하면서 개발 중입니다.
- 🤚🏻 질문 2 : Double 타입으로 생성한 lastModified 라는 Attribute 는 `옵셔널`이 체크되어 있음에도, 프로퍼티엔 옵셔널이 붙어서 나오지 않더라구요? Xcode 가 왜 이렇게 동작하는지 잘 모르겠습니다. 버그일까요...? 아니면 무슨 의도가 있는 것일까요? 🤔 

<p align="left"><img src="https://user-images.githubusercontent.com/71127966/154234509-118eb60d-81a0-4e19-b22d-e1e493766073.png" width="80%"></p>

<br>

### 2️⃣ CoreData context 생성과 CRUD 코드가 적절하게 구현됐을까요?

- 먼저, MemoStorage 라는 클래스 타입을 생성하고, 내부 프로퍼티로 NSPersistentContainer 를 만들고, 자주 호출될 `context` 또한 변수로 구현했습니다.
이때 [viewContext](https://developer.apple.com/documentation/coredata/nspersistentcontainer/1640622-viewcontext)를 사용하지 않고 [newBackgroundContext()](https://developer.apple.com/documentation/coredata/nspersistentcontainer/1640581-newbackgroundcontext)를 사용했습니다.
- 이유는 viewContext 는 `main queue`를 사용하지만, newBackgroundContext 메서드로 생성한 context 는 `private queue`를 따로 생성해서 사용하기 때문입니다. 굳이 메인 스레드를 사용하며 연산 비용을 높여주고 싶지 않았습니다.
- ✋🏻 질문 1 : viewContext 가 아니라 `newBackgroundContext` 를 사용한 건 좋은 선택일까요?
부작용은 없을까요? 현업에서는 어떤 context 를 사용하는지 궁금합니다!
- ✋🏻 질문 2 : MemoStorage 클래스 내부에 있는 `CRUD` 메서드들이 적절하게 구현됐는지 점검해주시면 감사하겠습니다! 🙏🏻 
- ✋🏻 질문 3 : 프로퍼티에 `id`를 만들어두긴 했으나, 제대로 활용하지 못했습니다.
indexPath 를 사용해서 메모를 추적하는 로직을 만드느라 고생을 많이했는데요. 😓 
`id`를 기준으로 메모 삭제, 선택된 indexPath 변경 등을 손쉽게 추적하는 방법이 있을까요? 아이디어나 읽을 거리라도 공유해주시면 감사하겠습니다...! 😍 

```swift
lazy var context = persistentContainer.newBackgroundContext()
private var persistentContainer: NSPersistentContainer = {
    let container = NSPersistentContainer(name: "CloudNotes")
    container.loadPersistentStores { _, error in
        if let error = error {
            fatalError("persistent stores 로드에 실패했습니다 : \(error)")
        }
    }
    return container
}()
```

<br>

### 3️⃣ delegate 패턴이 적절하게 구현됐을까요?

- 스텝2로 넘어오면서, `뷰컨 간의 데이터 전달 방식`을 기존 `메서드 주입`에서 -> `delegate 패턴`으로 리팩토링했습니다!
이 과정에서 `추상화`를 위한 프로토콜을 2개 구현하며 역할 분리를 시도했습니다.
- ✋🏻 질문 1 : delegate 패턴으로의 리팩토링이 잘 이루어졌는지, 개선 가능한 부분이 있는지 궁금합니다!
- ✋🏻 질문 2 : 프로토콜을 2개로 나눌 정도로 delegate 에 `많은 역할`을 위임해줬는데요. 이렇게 역할을 많이 부여해줘도 괜찮은지 궁금합니다. 🤔 

```swift
typealias MemoManageable = MemoSplitViewManageable & MemoStorageManageable

protocol MemoStorageManageable: AnyObject {
    var isMemoStorageEmpty: Bool { get }
    var memosCount: Int { get }
    
    func create()
    func fetchAll()
    func fetch(at indexPath: IndexPath) -> Memo
    func delete(at indexPath: IndexPath)
    func update(at indexPath: IndexPath, title: String, body: String)
}

protocol MemoSplitViewManageable: AnyObject {
    func showPrimaryView()
    func showSecondaryView(of indexPath: IndexPath)
    func presentShareActivity(at indexPath: IndexPath)
    func presentDeleteAlert(at indexPath: IndexPath)
    func reloadRow(at indexPath: IndexPath, title: String, body: String)
}
```

<br>

### 4️⃣ 메모를 작성하던 도중 앱이 갑자기 종료된다면, 변경 사항을 어떻게 CoreData 에 저장해줄 수 있을까요?

- 저희는 메모 textView 에 내용이 바뀔 때 마다 `update`하는 것은 연산 비용이 크다고 생각하여, `UITextViewDelegate` 프로토콜의 [textViewDidEndEditing](https://developer.apple.com/documentation/uikit/uitextviewdelegate/1618628-textviewdidendediting) 메서드를 사용해서, textView 가 `focus 를 잃을 때 update` 해주도록 구현했습니다.
- 근데, textView 에 메모를 작성하는 도중(focus 를 잃지 않은 상태에서) `앱을 강제 종료`하면, 작성하던 메모가 저장이 되지 않더라구요. 🥲 
- 해당 이슈를 해결하기 위해, 뷰컨이 `deinit`될 때 textViewDidEndEditing 메서드를 호출해서 update 로직을 타게 만드려고 했으나, 앱이 강제 종료되면 deinit 이 호출되지 않는다는 걸 발견했습니다.
- ✋🏻 질문 1 : 앱이 강제종료 될 때 인스턴스가 모두 `deinit` 될텐데, 그 내부의 코드가 동작하지 않는 이유가 무엇일까요?
deinit 내부 코드는 인스턴스가 메모리에서 해제되기 `이전`이 아니라 `이후`에 호출되는 건데, 앱이 통째로 메모리에서 이미 내려간 상태라, 코드를 실행해줄 주체(?)가 없기 때문일까요?
- ✋🏻 질문 2 : 그렇다면 앱이 종료되기 직전에 textViewDidEndEditing 메서드를 호출해주려면 어떻게 해야 할까요?
[applicationWillTerminate](https://developer.apple.com/documentation/uikit/uiapplicationdelegate/1623111-applicationwillterminate) 같은 `App life cycle`를 활용하면 될까요? 🤔 

```swift
extension MemoDetailViewController: UITextViewDelegate {
    func textViewDidEndEditing(_ textView: UITextView) {
        // textView 가 비어있는 상태로 focus 를 잃으면 -> 메모를 삭제합니다.
        guard textView.hasText else {
            delegate?.delete(at: currentIndexPath)
            return
        }
        
        // textView 변경사항이 없다면 -> update 를 호출하지 않습니다.
        guard currentText != textView.text else {
            return
        }
        
        // textView 변경사항이 있다면 -> title 과 body 로 분리하고 update 합니다.
        let (title, body) = splitText(text: textView.text)
        delegate?.update(at: currentIndexPath, title: title, body: body)
        
        currentIndexPath = .zero
    }
}
```

<br>

### 5️⃣ 앱의 전반적인 사용자 경험을 체크해주시면 감사하겠습니다!

- 세세한 기능이 많이 적용된 앱이라, PR이나 코드 만으로는 표현하기 어려운 부분이 많았습니다.
- cloning 을 하셔서, 코리가 앱을 실제로 사용해보며 불편하셨던 부분이나, 사용자 관점에서 추가로 꼭 구현해야 할 기능이 있는지 편하게 의견을 주실 수 있을까요? 😄 

---

감사합니다!

Jager-yoo committed:
feat: CoreData 를 활용한 MemoStorage 타입 구현

ChaminLee committed:
feat: MemoStorage내 create 메서드 구현

- MemoStorage struct -> class
- CloudNotes내 Entity 추가 및 subclass 생성
- Memo 타입내 id 프로퍼티 추가 및 Codable, CodingKeys 제거
- UnitTest 파일, JSONConverter 타입 제거
- configureTableView() 메서드 내 데이터가 없는 경우 crash 방지하는 조건 추가

Jager-yoo committed:
file: CloudNotes 의 Subclass 파일 추가

Jager-yoo committed:
refactor: 기존 Memo 타입을 삭제하고 CoreData 모델로 대체

ChaminLee committed:
refactor: MemoStorage create 메서드 로직 수정

Jager-yoo committed:
feat: MemoStorage 타입의 fetchAll 메서드 구현

- MemoTableViewController 가 MemoStorage 타입을 프로퍼티로 소유하도록 함
- memo 프로퍼티는 lazy 로 구현하여, 사용될 때 초기화되도록 함

ChaminLee committed:
feat: 새로운 메모 추가 액션 구현

- 메모를 추가한 후 tableView의 첫 번째 index에 추가/업데이트 해주는 addEmptyMemo 메서드 생성
- memo를 업데이트 하는 fetchMemo 메서드 생성
- 최상단의 indexPath를 initialIndexPath 프로퍼티에 저장
- memo의 title과 body가 공백일 때 placeholderText로 대체

Jager-yoo committed:
feat: Swipe 삭제 기능 구현

- MemoStorage 타입에 delete 메서드, saveContext 메서드 구현
- MemoTableViewController 타입에 deleteMemo 메서드 구현
- M1맥에서 framework not found Pods 에러가 발생하여 Frameworks 폴더 내에 있는 파일 삭제함

Jager-yoo committed:
refactor: UIContextualAction 타입의 deleteAction style 변경 normal -> destructive

ChaminLee committed:
feat: MemoDetailViewController내 메모 삭제 기능 구현

- MemoSplitViewController에서 메모를 삭제할 수 있도록 도와주는 deleteMemo 메서드 구현
- MemoTableViewController 내 삭제 의사를 묻는 alert을 띄우는 presentDeleteAlert 메서드 구현
- showSecondaryView 에 indexPath도 전달하도록 수정
- MemoDetailViewController가 현재 indexPath를 저장하도록 프로퍼티 추가

Jager-yoo committed:
feat: 멀티태스킹 모드일 때 Secondary view 에서 메모 삭제 시 Primary view 로 이동하는 기능 구현

ChaminLee committed:
feat: MemoDetailViewController내 더보기 버튼 pull-down 방식으로 구현

Jager-yoo committed:
feat: UIActivityViewController 이용한 메모 공유 기능 구현

ChaminLee committed:
refactor: ViewController 사이의 구조를 delegate 패턴을 활용하여 변경

- MemoSplitViewManageable 프로토콜을 통해 상위 뷰컨에 해당하는 memoSplitViewController에 요청을 보낼 수 있다.
- MemoStorageManageable 프로토콜을 통해 메모 데이터를 관리하고 하위 뷰컨트롤러에서 데이터를 받도록 해준다.
- 두 프로토콜을 typealias하여 MemoManageable이라고 네이밍

Jager-yoo committed:
refactor: MemoDetailVC 와 상위 VC 사이의 구조를 delegate 패턴을 활용하여 변경

ChaminLee committed:
feat: update 메서드 구현

- coreData 업데이트
- 메모의 제목과 내용이 비어있을 때, endEditing에서 확인하여 해당 메모를 삭제하도록 구현

ChaminLee committed:
feat: 메모 업데이트 기능 구현

- memo의 값이 변경될 때 마다 데이터가 정렬되도록 구현
- textview의 text가 변경될 때마다 cell의 label들을 업데이트하고, endEditing이 호출될 때 coreData에 최종 수정 완료된 데이터 저장 후 row 이동
- begin editing과 end editing을 비교하여 변경사항이 없다면 업데이트 하지 않도록 구현

Jager-yoo committed:
fix: 메모의 제목과 내용 사이 줄바꿈을 고려하여 내용의 첫 번째 줄이 preview 라벨에 들어가도록 수정

Jager-yoo committed:
feat: 메모 textView 의 제목과 내용을 줄바꿈으로 구분하여 폰트가 달라지는 기능 구현

ChaminLee committed:
fix: 메모 삭제에 따른 화면 전환 및 selectedIndexPath 업데이트

- 메모가 1개 남아있을 때는 삭제하지 못하도록 alert 노출
- selectedIndexPath와 삭제될 IndexPath를 비교하여 삭제 이후 selectedIndexPath를 업데이트 하는 로직 구현

Jager-yoo committed:
feat: 앱 최초 실행 시, 기본적으로 메모 1개가 저장되어 있도록 구현

- 앱을 처음 설치하면 CoreData 가 비어있는 상태이므로, 기본적으로 1개의 메모를 viewDidLoad 에서 미리 저장해두는 로직 추가
- SceneDelegate, AppDelegate 위치한 중복된 Core Data Stack 제거함

ChaminLee committed:
refactor: 상수 관리

corykim0829 commented on CloudNotes/CloudNotes/Controller/MemoSplitViewController.swift:
```diff
@@ -7,14 +7,48 @@
 
 import UIKit
 
+private enum AlertMessage {
```
private 으로 사용되기 때문에 nested enum으로 하는게 더 깔끔할 것 같아요!

ChaminLee committed:
fix: 텍스트를 입력할 때 마다 core data에 저장하도록 로직 수정

- 기존 endEditing에서 저장하던 방식에서 매번 입력될 때 마다 저장되도록 수정
- 새로운 메모를 생성할 때 항상 첫 번째 셀과 그 내용이 보이도록 구현
- 메모를 삭제할 때 선택된 셀, 지우고자 하는 셀의 indexPath를 비교하는 로직을 통해 제거
- 텍스트가 업데이트 될 때, performBatchUpdates를 통해 moveRow가 선행되고, reloadRows, selectRows가 호출되도록 구현

ChaminLee commented:
### 1. Core Data 업데이트 시점 textViewDidEndEditing -> `textViewDidChange`으로 수정했습니다!

기존에 위 PR에서 말씀드린 것 처럼 `textViewDidChange`에서 텍스트만 변경해주고, `textViewDidEndEditing`에서 실제로 core data에 접근하여 업데이트 해주도록 했었습니다. 다만 여러 상황에서 발생할 수 있는 에러들이 많아서 `textViewDidChange`에서 바로 core data의 데이터를 업데이트 해주도록 변경했습니다!

corykim0829 reviewed (approved):
정말 정말 수고 많으셨습니다!!

전반적으로 너무 깔끔하게 잘 구현하셔서 크게 리뷰 남길 부분이 없었던 것 같아요!

제가 남긴 리뷰와 질문에 대한 답변을 기반으로 수정하는 부분은 다음 스텝 진행하면서 같이 진행하면 될 것 같아요!

제가 답변한 부분에 대해서 궁금하신 점 있으면 언제든지 DM 주세요 :)

고생 너무 많으셨습니다!

---

1️⃣ CoreData 모델을 커스텀해서 사용해도 될까요?

➡️ Core Data에서 default로 옵셔널로 처리되는 이유는 Core Data와 Swift에서 처리되는게 달라서이기 떄문에 저도 개인적으로는 번거롭지만 옵셔널 처리로 처리하는게 맞다고 생각합니다.
Double 같이 Core Data 어트리뷰트 중 옵셔널이 아닌 값은 저장이 될 때 nil 값이 되지 않도록 처리되기 때문인걸로 알고 있어요. 버그는 아닐거에요 ㅎ


2️⃣ CoreData context 생성과 CRUD 코드가 적절하게 구현됐을까요?

➡️ newBackgroundContext 사용은 더 좋은 퍼포먼스를 기대할 수 있는 것 같아요 👍
CRUD도 모두 잘 구현하셨습니다 👏👏👏👏

`id`를 기준으로 메모 삭제, 선택된 indexPath 변경 등을 손쉽게 추적하는 방법이 있을까요? 
이 부분이 정확하게 어떤 의미인지 이해가 잘 안되네요!

3️⃣ delegate 패턴이 적절하게 구현됐을까요?

➡️ 프로토콜 구현 자체는 매우 좋습니다. 다만 프로토콜을 두개로 분리했지만 결국 MemoSplitVC의 하위 뷰컨들이 모두 MemoManageable 프로토콜을 채택을 해서 나눈 이유가 크게 없게 느껴지는 것 같아요.
정말 필요한 프로토콜만 채택할 정도로 잘 나누는 것도 좋을 것 같아요!
역할을 많이 부여해서 이 프로토콜을 잘 활용하면 아무 문제 없습니다 :)

4️⃣ 메모를 작성하던 도중 앱이 갑자기 종료된다면, 변경 사항을 어떻게 CoreData 에 저장해줄 수 있을까요?

➡️ 넵! App life cycle을 활용해서 처리하는게 좋을 것 같아요. deinit이 안되는 부분은 저도 잘 모르는 부분이라 저도 찾아봐야할 것 같네요! 알게되면 말씀드릴게요 ㅎㅎ

5️⃣ 앱의 전반적인 사용자 경험을 체크해주시면 감사하겠습니다!

➡️ 이전에 피드백 드린 부분도 모두 수정하셔서 아주 좋습니다 👍
